server/src/main/java/org/elasticsearch/index/mapper/PostParsePhase.java

```diff
@@ -0,0 +1,533 @@
+
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.DocValuesType;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.FieldInfos;
+import org.apache.lucene.index.Fields;
+import org.apache.lucene.index.IndexOptions;
+import org.apache.lucene.index.IndexableField;
+import org.apache.lucene.index.LeafMetaData;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.NumericDocValues;
+import org.apache.lucene.index.PointValues;
+import org.apache.lucene.index.SortedDocValues;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.index.StoredFieldVisitor;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.bytes.BytesReference;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Executes post parse phases on mappings
+ */
+public class PostParsePhase {
+
+    private final Map<String, OneTimeFieldExecutor> fieldExecutors = new HashMap<>();
+    private final PostParseContext context;
+
+    /**
+     * Given a mapping, collects all {@link PostParseExecutor}s and executes them
+     * @param lookup        the MappingLookup to collect executors from
+     * @param parseContext  the ParseContext of the current document
+     */
+    public static void executePostParsePhases(MappingLookup lookup, ParseContext parseContext) throws IOException {
+        PostParsePhase postParsePhase = lookup.buildPostParsePhase(parseContext);
+        if (postParsePhase == null) {
+            return;
+        }
+        postParsePhase.execute();
+    }
+
+    PostParsePhase(
+        Map<String, PostParseExecutor> postParseExecutors,
+        Function<String, MappedFieldType> fieldTypeLookup,
+        ParseContext pc) {
+        LazyDocumentReader reader = new LazyDocumentReader(
+            pc.rootDoc(),
+            pc.sourceToParse().source(),
+            postParseExecutors.keySet());
+        this.context = new PostParseContext(fieldTypeLookup, pc, reader.getContext());
+        postParseExecutors.forEach((k, c) -> fieldExecutors.put(k, new OneTimeFieldExecutor(c)));
+    }
+
+    void execute() throws IOException {
+        for (OneTimeFieldExecutor executor : fieldExecutors.values()) {
+            executor.execute();
+        }
+    }
+
+    // FieldExecutors can be called both by executePostParse() and from the lazy reader,
+    // so to ensure that we don't run field scripts multiple times we guard them with
+    // this one-time executor class
+    private class OneTimeFieldExecutor {
```

Comment:
this is effectively to deal with scripts that mention other scripted fields, so that that same script is not executed again for the same document?

Reply:
Correct